test/replication-luatest/gh_5295_split_brain_test.lua

```diff
@@ -0,0 +1,247 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('gh-6036')
+
+local function repl_list(i)
+    local repl = {}
+    for j = 1,i do
+        table.insert(repl, server.build_instance_uri('s'..j))
+    end
+    return repl
+end
+
+-- gh-5295: the server should stop replication from an upstream which sends data
+-- conflicting in terms of promote / confirmed_lsn.
+--
+-- The test requires one instance in replicaset per each split-brain case, plus
+-- one "main" instance. The idea is to partition one instance from the main
+-- server per test case. Once partitioned, create a split-brain situation
+-- between the partitioned node and all the main's followers. Then check that
+-- the partitioned node can't reconnect.
+local num_instances = 1
+
+g.before_all(function(cg)
+    cg.cluster = cluster:new({})
+
+    cg.box_cfg = {
+        replication = repl_list(num_instances),
+        replication_timeout         = 0.1,
+        replication_synchro_quorum  = 1,
+        replication_synchro_timeout = 0.01,
+        election_mode               = 'manual',
+        election_timeout            = 0.1,
+        election_fencing_enabled    = false,
+        log_level                   = 6,
+    }
+
+    cg.s = {}
+    for i = 1,num_instances do
+        cg.s[i] = cg.cluster:build_and_add_server({
+            alias = 's'..i,
+            box_cfg = cg.box_cfg,
+        })
+    end
+    cg.cluster:start()
+
+    cg.main_id = cg.s[1]:exec(function()
+        box.ctl.promote()
+        box.ctl.wait_rw()
+        local s = box.schema.space.create('sync', {is_sync = true})
+        s:create_index('pk')
+        s = box.schema.space.create('async')
+        s:create_index('pk')
+        return box.info.id
+    end)
+end)
+
+g.before_each(function(cg)
+    -- Check that the servers start synced and with s[1] being leader.
+    -- It's a prerequisite for each test.
+    cg.s[1]:exec(function()
+        require('luatest').assert_equals(box.info.synchro.queue.owner,
+                                         box.info.id, 'First node is leader')
+    end)
+    t.helpers.retrying({}, function()
+        cg.s[num_instances]:assert_follows_upstream(cg.main_id)
+    end)
```

Comment:
But you check it only for the last server: `cg.s[num_instances]`, don't you?

Reply:
Yes. I only care about the next testing server (`cg.s[num_instances]`) and its connection to the leader.
But the testing server changes before each test: there's `num_instances = num_instances - 1` in `g.after_each`.